rfd/0008-api.md

```diff
@@ -0,0 +1,241 @@
+---
+authors: Alexander Klizhentas (sasha@gravitational.com)
+state: discussion
+---
+
+# RFD 8 - API and Client Libraries
+
+## What
+
+Document and publish API OpenAPI spec and documentation.
+Unify Golang Client experience.
+
+## Why
+
+There are several problems with Go client libraries:
+
+* Go library example pulls entire teleport dependency including `lib/auth` and `lib/services`.
+* There is no separate client library with guarantees of compatibility  with specific version of Teleport.
+See examples [here](https://github.com/kubernetes/client-go#compatibility-matrix).
+*  It's impossible to generate library in any other language other than Go due to unclear API surface.
+*  Some client logic is residing in [plugins](https://github.com/gravitational/teleport-plugins/blob/master/access/access.go),
+while other is in `lib/auth`.
+
+* Code in `lib/auth` uses some concepts  unfamiliar to Go users:
+
+```go
+// connectClient establishes a gRPC connection to an auth server.
+func connectClient() (*auth.Client, error) {
+    tlsConfig, err := LoadTLSConfig("certs/api-admin.crt", "certs/api-admin.key", "certs/api-admin.cas")
+    if err != nil {
+        return nil, fmt.Errorf("Failed to setup TLS config: %v", err)
+    }
+
+    // must part addr list panics and is not necessary
+    authServerAddr := utils.MustParseAddrList("127.0.0.1:3025")
+    clientConfig := auth.ClientConfig{Addrs: authServerAddr, TLS: tlsConfig}
+
+    // TLS client is the only client
+    return auth.NewTLSClient(clientConfig)
+```
+
+## Details
+
+### Approach
+
+Keep Teleport's API layer unified, avoid having internal API vs external API layer.
+The access to the API surface is limited by role based access control.
+
+Avoid large refactors, instead use Go's type aliases to move things around while keeping things backwards compatible.
+
+```
+// Preserve full backwards compatibility while keeping
+type services.Role = types.Role
```

Comment:
nit: I don't think it is possible to create an alias for a type in another package: e.g. in `type T1 = T2`, T1 cannot be from another package iirc but should be an [identifier](https://golang.org/ref/spec#Identifiers).

Reply:
Was not aware that's not possible, if we are introducing breaking changes, may be we can create a migration tool using ast/parse to rename the imports and add to the changelog @Joerger

Reply:
It can be done like this:
```go
// ClientRole is a client alias type for teleport.Role
type ClientRole teleport.Role

const (
	// RoleAuth is for teleport auth server (authority, authentication and authorization)
	RoleAuth ClientRole = ClientRole(teleport.RoleAuth)
)
```